inference/server/oasst_inference_server/plugins/web_retriever/main.py

```diff
@@ -0,0 +1,253 @@
+from fastapi.responses import JSONResponse
+from loguru import logger
+
+# In total, the text + image links + prompts should be <= 2048
+CHAR_LIMIT = 1585  # TODO: increase these values after long-context support has been added
+IMAGES_CHAR_LIMIT = 300
+MAX_DOWNLOAD_SIZE = 4 * 1024 * 1024
+
+IMAGES_SUFIX = """, and I will also include images formatted like this:
+![](image url)
+"""
+
+folder_path = Path(__file__).parent
+
+app = FastAPI()
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+
+def extract_image_links(text: str):
+    image_pattern = r"https?://\S+\.(?:jpg|jpeg|png|gif|bmp|webp|svg)"
+    images = re.findall(image_pattern, text, flags=re.IGNORECASE | re.MULTILINE)
+    return images
+
+
+def detect_content_type(content: bytes) -> str:
+    if content.startswith(b"%PDF-"):
+        return "application/pdf"
+    elif (content).upper().startswith(b"<!DOCTYPE HTML") or content.startswith(b"<html"):
+        return "text/html"
+    elif content.startswith(b"{") or content.startswith(b"["):
+        try:
+            json.loads(content)
+            return "application/json"
+        except json.JSONDecodeError:
+            pass
+    elif content.startswith(b"---") or content.startswith(b"%YAML"):
+        try:
+            yaml.safe_load(content)
+            return "application/x-yaml"
+        except yaml.YAMLError:
+            pass
+
+    return "text/plain"
+
+
+def limit_image_count(images, max_chars=300):
+    limited_images = []
+    current_length = 0
+
+    for url in images:
+        # Add the length of "http:" if the URL starts with "//"
+        url_length = len("http:") + len(url) if url.startswith("//") else len(url)
+
+        if current_length + url_length > max_chars:
+            break
+
+        if url.startswith("//"):
+            limited_images.append(f"http:{url}")
+        else:
+            limited_images.append(url)
+
+        current_length += url_length
+
+    return limited_images
+
+
+def truncate_paragraphs(paragraphs, max_length):
+    truncated_paragraphs = []
+    current_length = 0
+
+    for paragraph in paragraphs:
+        if current_length + len(paragraph) <= max_length:
+            truncated_paragraphs.append(paragraph)
+            current_length += len(paragraph)
+        else:
+            remaining_length = max_length - current_length
+            truncated_paragraph = paragraph[:remaining_length]
+            truncated_paragraphs.append(truncated_paragraph)
+            break
+
+    return truncated_paragraphs
+
+
+@app.get("/get-url-content/", operation_id="getUrlContent", summary="It will return a web page's or pdf's content")
+async def get_url_content(url: str = Query(..., description="url to fetch content from")) -> Response:
+    try:
+        buffer = io.BytesIO()
+        encoding: str | None
+        async with httpx.AsyncClient(follow_redirects=True) as client:
```

Comment:
Is there a reason we need to use `httpx` over `aiohttp` here? The server already depends on `aiohttp` so we could reduce dependencies by using it, if possible